zio-cli/shared/src/main/scala/zio/cli/oauth2/OAuth2.scala

```diff
@@ -0,0 +1,259 @@
+      .map(ref =>
+        new OAuth2Token {
+          val refreshTokenNow = ref.updateZIO(refreshAccessToken)
+          val accessToken =
+            ref.updateAndGetZIO { acc =>
+              Clock.currentDateTime.flatMap { time =>
+                acc.expiresAt.map(_.isBefore(time)) match {
+                  // Expired
+                  case Some(true) => refreshAccessToken(acc)
+                  // No expiration or not yet expired
+                  case _ => ZIO.succeed(acc)
+                }
+              }
+            }
+        }
+      )
+
+  /**
+   * Refreshes access token, either by re-authorizing or by
+   * using refresh token, if available.
+   *
+   * @param accessToken original access token
+   */
+  def refreshAccessToken(accessToken: AccessToken): Task[AccessToken] =
+    accessToken.refreshToken.fold(authorize)(requestRefreshToken)
+
+  /**
+   * Sends to an authorization server request for authorization.
+   */
+  def requestAuthorization: Task[AuthorizationResponse] =
+    ZIO
+      .fromCompletableFuture(
+        client.sendAsync(
+          provider.authorizationRequest(scope),
+          HttpResponse.BodyHandlers.ofString()
+        )
+      )
+      .flatMap(response =>
+        if (response.statusCode() == 200) {
+          ZIO
+            .fromEither(provider.decodeAuthorizationResponse(response.body()))
+            .mapError(e => new Exception(s"Response of authorization request could not be decoded: $e"))
+        } else {
+          ZIO.fail(
+            new Exception(s"Authorization server returned error after authorization request: ${response.body()}")
+          )
+        }
+      )
+
+  def waitForAccessToken(response: AuthorizationResponse): Task[AccessToken] =
+    ZIO
+      .fromCompletableFuture(
+        client.sendAsync(
+          provider.accessTokenRequest(response),
+          HttpResponse.BodyHandlers.ofString()
+        )
+      )
+      .flatMap(response =>
+        ZIO.fromEither(
+          provider
+            .decodeAccessTokenResponse(response.body())
+            .left
+            .map(e => new Exception(s"Response of access token request could not be decoded: $e"))
+        )
+      )
+      .repeat(pollingSchedule(response.interval, response.expiresIn))
+      .flatMap(res => processResponse(res, None))
+
+  /**
+   * Creates ZIO Schedule for regular polling of authentication server
+   * until the response indicates success, failure or time has expired.
+   *
+   * Initial interval is `interval`. This interval can increase by
+   * `slow_down` requests from server. The schedule will terminate when
+   * `expiresIn` has elapsed.
+   *
+   * @see https://datatracker.ietf.org/doc/html/rfc8628#section-3.3
+   * @see https://datatracker.ietf.org/doc/html/rfc8628#section-3.5
+   *
+   * @param interval initial polling interval
+   * @param expiresIn maximum period
+   * @return the last response received before terminating
+   */
+  def pollingSchedule(
```

Comment:
For things like this that we probably don't want to expose in the public API, be sure to mark `private[cli]`.

Reply:
I hid the whole class, it's not supposed to be accessible from outside.